### PR TITLE
Implement POST /v3/packages/:guid/upload

### PR DESCRIPTION
**Pull request by akrishna90**
https://github.com/cloudfoundry/cf-crd-explorations/issues/13

### CODE REVIEW

File: cfshim/handlers/package_handler.go [hunk 1]
```diff
@@ -274,3 +285,159 @@ func updateDockerPackageResponse(formattedPackage *CFAPIPresenterPackageResource
...
+func (p *PackageHandler) getPackagesListFromQuery(queryParameters map[string][]string) ([]*appsv1alpha1.Package, error) {
+	var filter Filter = &filters.PackageFilter{
+		QueryParameters: queryParameters,
+	}
+
+	// Get all the CF Apps from K8s API store in AllPackages which contains Items: []App
+	AllPackages := &appsv1alpha1.PackageList{}
+	err := p.Client.List(context.Background(), AllPackages)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching app: %v", err)
```
Comment (heycait) [inline cfshim/handlers/package_handler.go:422]: Error message should probably say "packages"

Also, now that we're doing this a lot more.. can something more general be extracted into a shared function so we have a single `getResourceListFromQuery`?

File: cfshim/handlers/package_handler.go [hunk 2]
```diff
@@ -274,3 +285,159 @@ func updateDockerPackageResponse(formattedPackage *CFAPIPresenterPackageResource
...
+	pkg := packages[0]
+
+	packageBitsFile, _, err := r.FormFile("bits")
+	if err != nil {
+		ReturnFormattedError(w, 500, "ServerError", err.Error(), 10001)
+		return
+	}
+	defer packageBitsFile.Close()
+
+	tmpFile, err := ioutil.TempFile(os.TempDir(), fmt.Sprintf("package-%s-", packageGuid))
```
Comment (heycait) [inline cfshim/handlers/package_handler.go:311]: Does something else get appended to the package name besides the guid? possibly an extra "-" here.

File: cfshim/handlers/package_handler.go [hunk 3]
```diff
@@ -274,3 +285,159 @@ func updateDockerPackageResponse(formattedPackage *CFAPIPresenterPackageResource
...
+		ReturnFormattedError(w, 500, "ServerError", err.Error(), 10001)
+		return
+	}
+
+	// Close the packageBitsFile
+	if err := tmpFile.Close(); err != nil {
+		ReturnFormattedError(w, 500, "ServerError", err.Error(), 10001)
+		return
+	}
+
```
Comment (heycait) [inline cfshim/handlers/package_handler.go:330]: Can a comment be left here explaining the blank image and all the layers?

File: cfshim/handlers/package_handler.go [hunk 4]
```diff
@@ -274,3 +285,159 @@ func updateDockerPackageResponse(formattedPackage *CFAPIPresenterPackageResource
...
+	err = remote.Write(ref, image, remote.WithAuth(authenticator))
+	if err != nil {
+		ReturnFormattedError(w, 500, "ServerError", err.Error(), 10001)
+		return
+	}
+
+	updatedPkg := pkg.DeepCopy()
+	// TODO: Update Package CR spec on K8s
```
Comment (heycait) [inline cfshim/handlers/package_handler.go:369]: This TODO need to happen in this PR?

File: cfshim/handlers/package_handler.go [hunk 5]
```diff
@@ -274,3 +285,159 @@ func updateDockerPackageResponse(formattedPackage *CFAPIPresenterPackageResource
...
+	updatedPkg.Spec.Source = appsv1alpha1.PackageSource{
+		Registry: appsv1alpha1.Registry{
+			Image:            ref.Name(),
+			ImagePullSecrets: nil, // TODO: What goes here? Maybe take this secret name as a config?
```
Comment (heycait) [inline cfshim/handlers/package_handler.go:373]: I imagine this should be the registry credentials we need to push the image to this registry?

File: cfshim/handlers/package_handler.go [hunk 6]
```diff
@@ -274,3 +285,159 @@ func updateDockerPackageResponse(formattedPackage *CFAPIPresenterPackageResource
...
+		},
+	}
+	meta.SetStatusCondition(&updatedPkg.Status.Conditions, metav1.Condition{
```
Comment (heycait) [inline cfshim/handlers/package_handler.go:376]: What exactly is going on with the status condition stuff? Just seems like things are being set but not quite sure what is happening here.

File: cfshim/handlers/package_handler.go [hunk 7]
```diff
@@ -274,3 +285,159 @@ func updateDockerPackageResponse(formattedPackage *CFAPIPresenterPackageResource
...
+		Reason:  "Uploaded",
+		Message: "",
+	})
+	err = p.Client.Patch(ctx, updatedPkg, client.MergeFrom(pkg))
+	if err != nil {
+		ReturnFormattedError(w, 500, "ServerError", err.Error(), 10001)
+		return
+	}
+
+	// Probably punting on this until the GET /v3/packages/:guid endpoint is implemented
```
Comment (heycait) [inline cfshim/handlers/package_handler.go:400]: I think the GET package is already in main?

File: cfshim/handlers/package_handler.go [hunk 8]
```diff
@@ -274,3 +285,159 @@ func updateDockerPackageResponse(formattedPackage *CFAPIPresenterPackageResource
...
+	// NOTE : based on knowledge shared by Birdrock about CREATE writing to etcd and GET/LIST reading from cache, which\
+	//		  can result in cache miss should it be sufficient to return the updated object itself rather than\
+	// 		  invoking the GET/LIST?
+
```
Comment (heycait) [inline cfshim/handlers/package_handler.go:404]: Yes I believe so. @Birdrock ^

File: cfshim/handlers/package_handler.go (hunk 2 above)
Comment (akrishna90) [reply cfshim/handlers/package_handler.go]: fixed in commit - [ccf28d794b1efa76dcf13477f138ea59c1724afa](https://github.com/cloudfoundry/cf-crd-explorations/pull/41/commits/ccf28d794b1efa76dcf13477f138ea59c1724afa)

File: cfshim/handlers/package_handler.go (hunk 4 above)
Comment (akrishna90) [reply cfshim/handlers/package_handler.go]: The TODO was taken care, the comment had to be removed. fixed in commit - [ccf28d794b1efa76dcf13477f138ea59c1724afa](https://github.com/cloudfoundry/cf-crd-explorations/pull/41/commits/ccf28d794b1efa76dcf13477f138ea59c1724afa)

File: cfshim/handlers/package_handler.go (hunk 6 above)
Comment (akrishna90) [reply cfshim/handlers/package_handler.go:375]: Package status conditions are used to determine a Package state. function `derivePackageState` in the same file takes a list of conditions as input and outputs a state. More about mapping between conditions and states can be found here - https://miro.com/app/board/o9J_lFiI8CU=/?moveToWidget=3074457361338757508&cot=14

File: cfshim/handlers/package_handler.go (hunk 7 above)
Comment (akrishna90) [reply cfshim/handlers/package_handler.go]: Yes. fixed in commit - [ccf28d794b1efa76dcf13477f138ea59c1724afa](https://github.com/cloudfoundry/cf-crd-explorations/pull/41/commits/ccf28d794b1efa76dcf13477f138ea59c1724afa)

File: cfshim/handlers/package_handler.go (hunk 1 above)
Comment (akrishna90) [reply cfshim/handlers/package_handler.go]: Right. I have taken a shot at building a generic function `getMatchingResources`. I haven't used it anywhere yet. Let me  know your thoughts about it

File: cfshim/handlers/package_handler.go (hunk 3 above)
Comment (akrishna90) [reply cfshim/handlers/package_handler.go:330]: @tcdowney & @matt-royal - requesting your help in addressing this

File: cfshim/handlers/package_handler.go (hunk 5 above)
Comment (akrishna90) [reply cfshim/handlers/package_handler.go:372]: Yes, I believe so.

Current implementation - `PackageRegistryUsername`, `PackageRegistryPassword` & `PackageRegistryTagBase` are fed as ENV variables. 

How do we want to handle this? 
1. Create a secret object for the above by hand and pass it as a env variable.
2. Continue to pass the three values & handle creating secret object in the code. 

@matt-royal @tcdowney @Birdrock @gnovv thoughts?

Comment (Birdrock) [reply cfshim/handlers/package_handler.go:372]: My gut reaction is that this should be in a Secret (credentials) and ConfigMap (any other configuration).

Comment (akrishna90) [reply cfshim/handlers/package_handler.go:372]: Do we need to change the implementation to read it from a configmap / secret ? Or is it sufficient to call it out here and leave the existing implementation as is?

File: cfshim/handlers/shared_functions.go [hunk 9]
```diff
@@ -49,3 +48,65 @@ func getAppListFromQuery(c *client.Client, queryParameters map[string][]string)
 	}
 	return matchedApps, nil
 }
+
+func getPackagesListFromQuery(c *client.Client, queryParameters map[string][]string) ([]*appsv1alpha1.Package, error) {
```
Comment (heycait) [inline cfshim/handlers/shared_functions.go:52]: You wouldn't need this one anymore with the generic below though right?

File: cfshim/handlers/package_handler.go (hunk 5 above)
Comment (tcdowney) [reply cfshim/handlers/package_handler.go:372]: > Continue to pass the three values & handle creating secret object in the code.

For `docker` type packages clients are able to specify username and password on a per-Package basis. If we want to support this use case I think we'll have to own the Secret creation here.

### docker package request

```json
curl "https://api.example.org/v3/packages" \
  -X POST \
  -H "Authorization: bearer [token]" \
  -H "Content-type: application/json" \
  -d '{
    "type": "docker",
    "relationships": {
      "app": {
        "data": {
          "guid": "[guid]"
        }
      }
    },
    "data": {
      "image": "registry/image:latest",
      "username": "username",
      "password": "password"
    }
  }'
```

### bits package request

```json
curl "https://api.example.org/v3/packages" \
  -X POST \
  -H "Authorization: bearer [token]" \
  -H "Content-type: application/json" \
  -d '{
    "type": "bits",
    "relationships": {
      "app": {
        "data": {
          "guid": "[guid]"
        }
      }
    }
  }'
```

For `bits` type Packages the management of the source code was always handled by the platform and users did not have the ability to provide credentials on a per Package basis. For this case I think we could get away with a single `Secret` that is configured at deploy time and have the Secret name be provided to the shim via an environment variable.

Comment (akrishna90) [reply cfshim/handlers/package_handler.go:372]: > For `bits` case I think we could get away with a single Secret that is configured at deploy time and have the Secret name be provided to the shim via an environment variable.

this approach will lock our shim to a single registry at any given time.

**Comment by matt-royal** [comment]
We should also squash or just reword the WIP commits in this PR. My intention was to deal with those before making the PR, but I was out sick and didn't mention it to those who took over.

File: cfshim/handlers/package_handler.go (hunk 3 above)
Comment (matt-royal) [reply cfshim/handlers/package_handler.go:330]: @heycait The interface of the image and the layers forces us to make a blank image, make a single layer and then append the layer to the image. It's definitely a little clunky, but the structure of the code seems pretty clear to me. Do you have any suggestions for what kind of comment would be helpful here?

**Comment by matt-royal** [comment]
@gnovv and I just accepted this as is. I decided against undoing the WIPs because it was causing conflicts. Should we just merge this PR, or do people have unresolved concerns?

File: cfshim/handlers/shared_functions.go (hunk 9 above)
Comment (akrishna90) [reply cfshim/handlers/shared_functions.go:52]: Right! the generic function doesn't provide a great deal of optimization in my view. I'm split in terms of which one to use. I'll leave them here as it doesn't break things.

**Comment by akrishna90** [comment]
> @gnovv and I just accepted this as is. I decided against undoing the WIPs because it was causing conflicts. Should we just merge this PR, or do people have unresolved concerns?

Thanks for updating the `README` with environment variable details. This is good to be merged.

**Pull request merged**